proto enums: fallback to 0-index value if present

viliusl opened the pull request:
As per [proto3 spec](https://developers.google.com/protocol-buffers/docs/proto3#default)  - `For enums, the default value is the first defined enum value, which must be 0.`

I added fallback to `0`-index value if present or previous behavior if not.

thesamet commented:
Hi @viliusl , thanks for the PR. The spec says

> When a message is parsed, if the encoded message does not contain a particular singular element, the corresponding field in the parsed object is set to the default value for that field

It doesn't define the behavior of when the value that is present is invalid. Looking at [JsonFormat as the reference implemention](https://github.com/protocolbuffers/protobuf/blob/master/java/util/src/main/java/com/google/protobuf/util/JsonFormat.java#L1906-L1931), we see that unrecognized strings should throw an exception (like we do already). It looks like we have a gap that unrecognized numeric values should become unrecognized enums. It looks like there is a flag called "ignoreUnknownFields" which may have the behavior you are looking for. If so, we can add a similar flag to our json parser, and add tests that show that the behavior matches Java's in all cases (where the flag is true/false, and when we pass unrecognized string and int values).

viliusl commented:
Thanks @thesamet for a swift response!

Now I adjusted implementation by:
 - using built-in `Enum.Unrecognized(idx)` instead of using 0 index for lookup. Imgo it should provide better compatibility with java impl;
 - left behaviour for unknown string enum value parsing (throws);
 - added support for `ignoringUnknownFields` flag.

Maybe I did smth wrong, but I expected for [test](https://github.com/scalapb/scalapb-json4s/pull/60/files#diff-ce830ec4bf9baebabfe9f5a88e6b067dR362) to result in `None`, whereas it resulted in `Some(MyEnum.UNKNOWN)`. Any ideas?

thesamet commented on src/main/scala/scalapb/json4s/JsonFormat.scala:
```diff
@@ -385,13 +385,15 @@ class Parser private (config: Parser.ParserConfig) {
-        s"Unexpected value ($value) for enum ${enumDescriptor.fullName}")
-  }
+      case JInt(v) => enumDescriptor.findValueByNumber(v.toInt)
+          .getOrElse(enumDescriptor.findValueByNumberCreatingIfUnknown(0))
```
shouldn't this be `findValueByNumberCreatingIfUnknown(v.toInt)`?

thesamet commented on src/main/scala/scalapb/json4s/JsonFormat.scala:
```diff
@@ -385,13 +385,15 @@ class Parser private (config: Parser.ParserConfig) {
+      case JInt(v) => enumDescriptor.findValueByNumber(v.toInt)
+          .getOrElse(enumDescriptor.findValueByNumberCreatingIfUnknown(0))
+      case JString(s) if config.isIgnoringUnknownFields => enumDescriptor.values.find(_.name == s)
+          .getOrElse(NullValue.NULL_VALUE.scalaValueDescriptor)
```
This is probably what causing the unexpected test outcome. `NullValue` is unrelated to the arbitrary enum we have here. The tricky part is to find a way to bubble up `None` out of `defaultEnumParser` - maybe the signature needs to chage to return `Option[EnumValueDescriptor]`, and that would require a binary-incompatible change to [registerEnumFormatter](`https://github.com/scalapb/scalapb-json4s/blob/7277b4825ad12e661767f77451f7f1b8f51b3a97/src/main/scala/scalapb/json4s/JsonFormat.scala#L43). If we go this path, then this PR would go on the next minor version due to the incompatibility.

viliusl commented:
@thesamet - yes, it's a binary incompatability, but imho impl now matches that of java. wdyt?

thesamet commented on src/main/scala/scalapb/json4s/JsonFormat.scala:
```diff
@@ -384,21 +388,26 @@ class Parser private (config: Parser.ParserConfig) {
+      case JInt(v) => enumDescriptor.findValueByNumber(v.toInt)
+          .orElse(Some(enumDescriptor.findValueByNumberCreatingIfUnknown(v.toInt)))
+      case JString(s) if config.isIgnoringUnknownFields => enumDescriptor.values.find(_.name == s)
+          .orElse(Some(NullValue.NULL_VALUE.scalaValueDescriptor))
```
this needs to throw for java compat?

viliusl replied:
You are right. Sorry - did all changes for this one and then left it out:/

viliusl replied:
Ups. Did all the changes for this one and left if `git stash`ed somewhere. Sorry. Fixed.

thesamet commented on src/test/scala/scalapb/json4s/JsonFormatSpec.scala:
```diff
@@ -343,6 +345,25 @@ class JsonFormatSpec extends FlatSpec with MustMatchers with OptionValues with J
+
+  "TestProto" should "not fail for unknown enum values when `ignoringUnknownFields` is set" in {
+      new Parser().ignoringUnknownFields
+        .fromJsonString[MyTest]("""{"opt_enum":"ZAZA"}""")  must be(MyTest(optEnum = Some(MyEnum.UNKNOWN)))
```
This should be None

thesamet commented on src/main/scala/scalapb/json4s/JsonFormat.scala:
```diff
@@ -27,9 +27,13 @@ case class JsonFormatException(msg: String, cause: Exception) extends Exception(
 case class FormatRegistry(
   messageFormatters: Map[Class[_], Formatter[_]] = Map.empty,
-  enumFormatters: Map[EnumDescriptor, Formatter[EnumValueDescriptor]] = Map.empty,
+  enumFormatters: Map[EnumDescriptor, EnumFormatter[EnumValueDescriptor]] = Map.empty,
```
Can you add a `@deprecated def this()` constructor that takes enumFormatters in the old signature and calls into this constructor? This way we will not break source compatibility.

viliusl replied:
I don't think I can do this because of default args. Will the fact that `old` version of `FormatRegistry` will require `new` to be created is an issue as well? Also what about `registerEnumFormatter` ?

I'm not really familiar with `source compatibility`, so need your guidance here.

thesamet commented:
I'm ok with the binary incompatiblity, we'll bump minor version. Please add tests that demonstrate identical java behavior (see what's avaiable in JavaAssertions.scala which is extended by JsonFormatSpec)

thesamet commented:
and of course, thanks for working on this PR!

viliusl commented:
`Please add tests that demonstrate identical java behavior` - added compat tests. test for enum with non-existing enum value turned-out funny, but did not figure-out a better way to test it. Could not test it via 
```
assertJsonIsSameAsJava(jsontest.test.EnumTest(Some(MyEnum.Unrecognized(10))))
```
due to special assert in scalapb generated code.

thesamet commented on src/main/scala/scalapb/json4s/JsonFormat.scala:
```diff
@@ -384,21 +387,25 @@ class Parser private (config: Parser.ParserConfig) {
-      })
+      }
+
+      res.map(v => PEnum(v)).getOrElse(PEmpty)
```
nit: `fold(PEmpty)(PEnum(_))`

thesamet commented on src/test/scala/scalapb/json4s/EnumFormatSpec.scala:
```diff
@@ -0,0 +1,33 @@
+    )
+  }
+
+  "Enum" should "be serialized the same way as java - non-existent int value" in {
```
"be parsed"

thesamet commented on src/test/scala/scalapb/json4s/EnumFormatSpec.scala:
```diff
@@ -0,0 +1,33 @@
+    val parser = JavaJsonParser.ignoringUnknownFields()
+    val enumJson = """{"enum": 10}"""
+
+    javaParse(enumJson, jsontest.Test.EnumTest.newBuilder, parser).toString must be("")
```
It looks like fact that this test pass shows a difference in behavior. The java parser lost the unrecognized value.

thesamet replied:
A good waqy to show that you are getting the same object: https://github.com/scalapb/scalapb-json4s/blob/master/src/test/scala/scalapb/json4s/JsonFormatSpec.scala#L202-L206 is to convert enumtest from java to scala and assert equality.

viliusl replied:
Used your advice for testing java/scala interop - [2 tests from here](https://github.com/viliusl/scalapb-json4s/blob/0fb8bc5503fd874dcafb52e826a001ed59694be9/src/test/scala/scalapb/json4s/EnumFormatSpec.scala#L41-L40)

I also extended tests with explicit proto2/proto3 variations - which led me to [adding this branch](https://github.com/viliusl/scalapb-json4s/blob/53bf40ad71848e71c3d276cf9bb7a079904a33f6/src/main/scala/scalapb/json4s/JsonFormat.scala#L393-L392)

There isssue though with Scala/Java compatability:
 - in java, (proto2/proto3), for unknown string enum value and `ignoringUnknownFields` you get value set to null;
 - in scala proto2 - you get set to `None`;
 - in scala proto3 - enum field is not `Optional`, so I cannot set it to `None`. I did set it to `0-index` value as per proto3 spec, but it's not symmetrical to java

what would be your verdict on this behavior diff?

thesamet replied:
Can you double-check the behavior in java where you are getting nulls? I wasn't able to reproduce this with protobuf-3.8.0, and if this is true, you have probably found a bug in protobuf-java. Here is what I tried:

    import jsontest.{Test, Test3}
    val p2Builder = Test.TestAllTypes.newBuilder()
    JavaJsonParser.ignoringUnknownFields().merge("""{"optionalNestedEnum": "TTT"}""", p2Builder)
    val p2 = p2Builder.build()
    println("P2", p2)
    println("P2", p2.getOptionalNestedEnum)

    val p3Builder = Test3.MyTest3.newBuilder()
    JavaJsonParser.ignoringUnknownFields().merge("""{"optEnum": "TTT"}""", p2Builder)
    val p3 = p3Builder.build()
    println("P3", p3)
    println("P3", p3.getOptEnum)

In both cases I am getting default values.

thesamet commented on src/test/scala/scalapb/json4s/JsonFormatSpec.scala:
```diff
@@ -343,6 +345,25 @@ class JsonFormatSpec extends FlatSpec with MustMatchers with OptionValues with J
   }
 
+  "TestProto" should "infer default value when present" in {
+    new Parser().fromJsonString[MyTest]("""{"opt_enum":10}""") must be(MyTest(optEnum = Some(MyEnum.Unrecognized(10))))
```
Can you move to the new EnumFormatSpec? Also add java-equivalence tests.

viliusl replied:
Moved new Enum tests to EnumFormatSpec. Did not add java/scala failure symmetricity tests. Should I?

thesamet replied:
Yes, the parsers should accept and reject the same inputs. Thank you!

viliusl replied:
If you feel tests are not really clear let me know - I can try to re-factor them. Now I think coverage is good, just specifying all proto2/3 ignoringUnknownFields true/false variations is a bit tricky.

thesamet commented:
See above for a code fragment that verifies that the parsed objects are equivalent.

viliusl commented:
@thesamet - please see my comment on `There isssue though with Scala/Java compatability`.

viliusl commented:
@thesamet - I did my changes. wdyt?

thesamet commented:
@viliusl , thank you for your PR. Your work has been merged via 72ce92a
I found a few additional cases where Java had a behavior that mismatched Scala: when a value was provided as a string (`"1"`), or as a decimal `1.0`. I also made the Java/Scala comparison function a little easier to use https://github.com/scalapb/scalapb-json4s/blob/master/src/test/scala/scalapb/json4s/EnumFormatSpec.scala

viliusl commented:
@thesamet thanks alot. Any eta for release?

thesamet commented:
This was released in 0.10.0
